cli_tools/common/utils/storage/resource_location_retriever.go

```diff
@@ -0,0 +1,150 @@
+// If no storageRegion is provided, GCE Zone from the running process
+// will be used.
+func (rlr *ResourceLocationRetriever) GetZone(storageLocation string, project string) (string, error) {
+	zone := ""
+	var err error
+	if storageLocation != "" {
+		// pick a zone from the region where data is stored
+		zone, err = rlr.getZoneFromStorageLocation(storageLocation, project)
+		if err == nil {
+			return zone, err
+		}
+	}
+
+	// determine zone based on the zone Cloud Build is running in
+	if rlr.Mgce.OnGCE() {
+		zone, err = rlr.Mgce.Zone()
+	}
+	if err != nil {
+		return "", daisy.Errf("can't infer zone: %v", err)
+	}
+	if zone == "" {
+		return "", daisy.Errf("zone is empty")
+	}
+	fmt.Printf("[image-import] Zone not provided, using %v\n", zone)
+
+	return zone, nil
+}
+
+func (rlr *ResourceLocationRetriever) getZoneFromStorageLocation(location string, project string) (string, daisy.DError) {
+	if project == "" {
+		return "", daisy.Errf("project cannot be empty in order to find a zone from a location")
+	}
+	zones, err := rlr.ComputeGCEService.ListZones(project)
+	if err != nil {
+		return "", daisy.Errf("Failed to list zones: %v", err)
+	}
+	if rlr.isMultiRegion(location) {
+		return rlr.getBestZoneForMultiRegion(location, zones)
+	}
+	return rlr.getZoneForRegion(location, zones)
+}
+
+func (rlr *ResourceLocationRetriever) getZoneForRegion(region string, zones []*compute.Zone) (string, daisy.DError) {
+	for _, zone := range zones {
+		if isZoneUp(zone) && strings.HasSuffix(strings.ToLower(zone.Region), strings.ToLower(region)) {
+			return zone.Name, nil
+		}
+	}
+	return "", daisy.Errf("no zone found for %v region", region)
+}
+
+func (rlr *ResourceLocationRetriever) getMultiRegionForRegion(region string) string {
+	for multiRegion, regions := range multiRegions {
+		for _, aRegion := range regions {
+			if strings.EqualFold(aRegion, region) {
+				return multiRegion
+			}
+		}
+	}
+
+	return ""
+}
+
+func (rlr *ResourceLocationRetriever) getBestZoneForMultiRegion(multiRegion string, zones []*compute.Zone) (string, daisy.DError) {
+	for _, region := range multiRegions[multiRegion] {
+		if zone, err := rlr.getZoneForRegion(region, zones); err == nil {
+			return zone, nil
+		}
+	}
+	return "", daisy.Errf("no zones found for %v multi region", multiRegion)
+}
+
+func (rlr *ResourceLocationRetriever) isMultiRegion(location string) bool {
+	_, containsKey := multiRegions[strings.ToUpper(location)]
+	return containsKey
+}
+
+func isZoneUp(zone *compute.Zone) bool {
+	return zone != nil && zone.Status == "UP"
+}
+
+// GetLargestStorageLocation returns the largest storage location that includes provided argument.
+// If argument is a multi-region, the argument is returned. If argument is a region within a multi-region,
+// the multi-region is returned. If argument is a region not within a multi-region, argument is returned.
+func (rlr *ResourceLocationRetriever) GetLargestStorageLocation(storageLocation string) string {
```

Comment:
The meaning of 'Largest' is not straight forward to me. Is there a better word? Or a longer description.

Reply:
What about 'broadest'?